lib/iris/cube.py

```diff
@@ -1727,57 +1687,17 @@ def data(self):
             (10, 20)
 
         """
-        if self.has_lazy_data():
-            try:
-                # Realise the data, convert from a NaN array to a masked array,
-                # and if appropriate cast to the specified cube result dtype.
-                result = as_concrete_data(self._dask_array,
-                                          nans_replacement=ma.masked,
-                                          result_dtype=self.dtype)
-                self._numpy_array = result
-                self.dtype = None
-
-            except MemoryError:
-                msg = "Failed to create the cube's data as there was not" \
-                      " enough memory available.\n" \
-                      "The array shape would have been {0!r} and the data" \
-                      " type {1}.\n" \
-                      "Consider freeing up variables or indexing the cube" \
-                      " before getting its data."
-                msg = msg.format(self.shape, self.dtype)
-                raise MemoryError(msg)
-        return self._numpy_array
+        return self._data_manager.data
 
     @data.setter
-    def data(self, value):
-        if not (hasattr(value, 'shape') and hasattr(value, 'dtype')):
-            value = np.asanyarray(value)
-
-        if self.core_data is not None and self.shape != value.shape:
-            # The _ONLY_ data reshape permitted is converting a 0-dimensional
-            # array i.e. self.shape == () into a 1-dimensional array of length
-            # one i.e. data.shape == (1,)
-            if self.shape or value.shape != (1,):
-                raise ValueError('Require cube data with shape %r, got '
-                                 '%r.' % (self.shape, value.shape))
-
-        # Set lazy or real data, and reset the other.
-        if is_lazy_data(value):
-            self._dask_array = value
-            self._numpy_array = None
-        else:
-            if not ma.isMaskedArray(value):
-                # Coerce input to ndarray (including ndarray subclasses).
-                value = np.asarray(value)
-            self._numpy_array = value
-            self._dask_array = None
-
-        # Cancel any 'realisation' datatype conversion, and fill value.
-        self.dtype = None
-        self.fill_value = None
+    def data(self, data):
+        # Assign the new data to the data manager.
+        self._data_manager.data = data
+        # Clear the cube fill-value.
+        self._fill_value = None
```

Comment:
Although, just considering the real masked array case, rather than the lazy case (which I'm been focused on), you're right that it might make sense to set the `Cube.fill_value` from the masked array `fill_value` ... I'll see how this plays ...